### PR TITLE
Add time to EnSight and OpenFOAM reader classes, add PVD reader class

### PR DESCRIPTION
**Pull request by MatthewFlamm**
### Overview
#1536 needs to be merged first.  I'm pushing this now in case it helps frame some of the decisions in #1536 or for reviewing in that PR.  I can also combine these PR's but this would result in a very, very large PR like this one looks like currently.

This PR is a followon to #1536 that adds time support to existing EnSight and OpenFOAM readers.  A from-scratch PVD reader is also added.

This is the functionality that I am most excited to finally get into pyvista!  It can replace my hacked mess for reading in these files for analysis.

### Details

EnSight and OpenFOAM both handle time a little differently, so an abstract class is inherited that ensures that we have a consistent user interface for interacting with the reader for time handling.

Again, I tested the OpenFOAM reader against an internal dataset.  There are some quirks.  For example, a reader cannot be reused, but overall it works OK.  This is untested here.

Closes: #414 

TODO:

- [x] Keep up to date with Review in #1536
- [x] Add better docstrings to time classes
- [x] Upload PVD datasets to vtk-data
- [x] Add another test for PVD reader including one dataset that includes multiple parts
- [x] Improve test coverage

### CODE REVIEW

**Comment by MatthewFlamm** [comment]
I cannot reproduce the CI error in the example. I can directly run the example successfully locally. I suspect it may be a vtk version issue related to ordering of plotter operations.  At least this is the only difference I can see with respect to the existing gif example.

I also cannot reproduce the CI mypy at all.  Here is what I've tried:

```bash
git clone
cd pyvista
python3 -m venv venv  # python3.8.2
source venv/bin/activate
pip install -e .
pip install -r requirements_static.txt
make mypy
```

```txt
Running mypy static type checking
mypy pyvista/core/ --no-incremental
pyvista/core/imaging.py:16: error: Incompatible default for argument "output_type" (default has type "Type[floating[Any]]", argument has type "dtype[Any]")
pyvista/core/dataset.py:589: error: Result type of + incompatible in assignment
pyvista/core/dataset.py:617: error: Result type of * incompatible in assignment
pyvista/core/dataset.py:911: error: Incompatible return value type (got "Optional[ndarray[Any, Any]]", expected "ndarray[Any, Any]")
```

Not only do I not see the error on the CI, but I get other errors not on the CI.

Since these are both related to the PVD addition, if I cannot remedy this, I may remove that from this PR and open a new one for that piece later.

**Comment by akaszynski** [comment]
@MatthewFlamm, I (accidentally) pushed to main the type check changes that will let you run the type check locally with no issues.

**Comment by akaszynski** [comment]
> Now this PR is running into numpydoc failure "GL08: The object does not have a docstring" for the delibrately skipped docstrings. I don't see a way to disabe numpydoc checks on a per docstring basis. I think we either need to turn off GL08 check or duplicate the doscstrings.

You can disable it here:
https://github.com/pyvista/pyvista/blob/6c42ad31d3c7bdc4f1724e3bec69a150a2eef042/doc/conf.py#L93-L94

**Comment by akaszynski** [comment]
Still looking into this locally.  It's weird it's not skipping.  I'll fix this since I've made this a pain for PRs...

File: doc/conf.py [hunk 1]
```diff
@@ -125,6 +125,7 @@
 
     # classes inherit from BaseReader
     r'\.*Reader$',
+    # r'\.*OpenFOAMReader*',
```
Comment (akaszynski) [inline doc/conf.py:128]:
```suggestion
    r'\.*Reader\.*',
```

**Comment by akaszynski** [comment]
Fixed it and it passes now.  #1627 will take care of printing out all the warnings rather than the first one.

File: pyvista/utilities/reader.py [hunk 2]
```diff
@@ -647,6 +772,174 @@ class BinaryMarchingCubesReader(BaseReader):
...
 
 
+@dataclass(order=True)
+class PVDDataSet:
+    """Class for storing dataset info from PVD file."""
+
+    time: float
+    part: int
+    filename: str
+    group: str
```
Comment (akaszynski) [inline pyvista/utilities/reader.py:782]: Cool usage of `dataclass`!

File: pyvista/utilities/reader.py [hunk 3]
```diff
@@ -647,6 +772,174 @@ class BinaryMarchingCubesReader(BaseReader):
...
+        self._active_readers = []
+        self._time_values = []
+
+        self._set_filename(filename)
+        
+    @property
+    def reader(self):
+        """Return the PVDReader.
+        
+        ..note::
```
Comment (akaszynski) [inline pyvista/utilities/reader.py:823]:
```suggestion
        .. note::
```

Comment (akaszynski) [reply pyvista/utilities/reader.py:823]: This won't render correctly unless there's a space between `..` and `note`.

File: pyvista/utilities/reader.py [hunk 4]
```diff
@@ -647,6 +772,174 @@ class BinaryMarchingCubesReader(BaseReader):
...
+        """
+        return self._datasets
+
+    @property
+    def active_datasets(self):
+        """Return all active datasets.
+        
+        Returns
+        -------
+        list[:class:`pyvista.PVDDataSet`]
```
Comment (akaszynski) [inline pyvista/utilities/reader.py:857]: Thanks to using `numpydoc`, we don't have to use the `class` directive anymore when writing returns.

```suggestion
        list[pyvista.PVDDataSet]
```

File: pyvista/utilities/reader.py [hunk 5]
```diff
@@ -647,6 +772,174 @@ class BinaryMarchingCubesReader(BaseReader):
...
+        """
+        return self._active_readers
+
+    @property
+    def datasets(self):
+        """Return all datasets.
+        
+        Returns
+        -------
+        list[:class:`pyvista.PVDDataSet`]
```
Comment (akaszynski) [inline pyvista/utilities/reader.py:846]:
```suggestion
        list[pyvista.PVDDataSet]
```

File: pyvista/utilities/reader.py [hunk 6]
```diff
@@ -647,6 +772,174 @@ class BinaryMarchingCubesReader(BaseReader):
...
+        """
+        return self
+
+    @property
+    def active_readers(self):
+        """Return the active readers.
+        
+        Returns
+        -------
+        list[:class:`pyvista.BaseReader`]
```
Comment (akaszynski) [inline pyvista/utilities/reader.py:835]:
```suggestion
        list[pyvista.BaseReader]
```

File: pyvista/utilities/reader.py [hunk 7]
```diff
@@ -647,6 +772,174 @@ class BinaryMarchingCubesReader(BaseReader):
...
+        self._update_information()
+
+    def read(self):
+        """Read data from PVD timepoint.
+
+        Overrides :func:`pyvista.BaseReader.read`.
+
+        Returns
+        -------
+        :class:`pyvista.MultiBlock`
```
Comment (akaszynski) [inline pyvista/utilities/reader.py:877]:
```suggestion
        pyvista.MultiBlock
```

File: tests/utilities/test_reader.py [hunk 8]
```diff
@@ -178,6 +178,30 @@ def test_ensightreader():
...
+    reader.set_active_time_value(3.0)
+    assert reader.active_time_value == 3.0
+    mesh_3 = reader.read()
+
+    # assert all the data is different
+    for m_1, m_3 in zip(mesh_1, mesh_3):
+        assert not all(m_1['DENS'] == m_3['DENS'])
+
+    reader.set_active_time_point(0)
+    assert reader.active_time_value == 1.0
```
Comment (akaszynski) [inline tests/utilities/test_reader.py:204]:
```suggestion
    assert reader.active_time_value == 1.0

```

**Review by akaszynski — APPROVED** [review]
Some minor changes, but otherwise good to go.

I'm assuming that you're ready for review since the task list is complete.

**Comment by MatthewFlamm** [comment]
Yes, I've been trying to use "draft PR" for when it is still very much a WIP but feedback would be welcome.  And then task lists for when the majority of the code is finalized but there are some items to complete.  It doesn't always go according to plan.

**Pull request merged**